### PR TITLE
fix/circuit-order: Reliable order of circuits in proofs

### PR DESCRIPTION
**Pull request by naure**
_Extracted from #368_

### CODE REVIEW

File: ceno_zkvm/src/scheme/tests.rs [hunk 1]
```diff
@@ -120,7 +120,12 @@ fn test_rw_lk_expression_combination() {
         // get proof
         let prover = ZKVMProver::new(pk);
         let mut transcript = Transcript::new(b"test");
-        let wits_in = zkvm_witness.witnesses.remove(&name).unwrap().into_mles();
+        let wits_in = zkvm_witness
+            .into_iter_sorted()
+            .next()
```
Comment (matthiasgoergens) [inline ceno_zkvm/src/scheme/tests.rs:125]: Isn't that just `first`?

Comment (naure) [reply ceno_zkvm/src/scheme/tests.rs:125]: That somehow didn't work.

File: ceno_zkvm/src/structs.rs [hunk 2]
```diff
@@ -205,7 +205,8 @@ impl<E: ExtensionField> ZKVMFixedTraces<E> {
 
 #[derive(Default)]
 pub struct ZKVMWitnesses<E: ExtensionField> {
-    pub witnesses: BTreeMap<String, RowMajorMatrix<E::BaseField>>,
+    witnesses_opcodes: BTreeMap<String, RowMajorMatrix<E::BaseField>>,
```
Comment (matthiasgoergens) [inline ceno_zkvm/src/structs.rs:208]: Wouldn't we want the kinds instead of the opcodes?

File: ceno_zkvm/src/structs.rs [hunk 3]
```diff
@@ -273,10 +275,18 @@ impl<E: ExtensionField> ZKVMWitnesses<E> {
...
-        assert!(self.witnesses.insert(TC::name(), witness).is_none());
+        assert!(self.witnesses_tables.insert(TC::name(), witness).is_none());
+        assert!(!self.witnesses_opcodes.contains_key(&TC::name()));
 
         Ok(())
     }
+
+    /// Iterate opcode circuits, then table circuits, sorted by name.
+    pub fn into_iter_sorted(self) -> impl Iterator<Item = (String, RowMajorMatrix<E::BaseField>)> {
+        self.witnesses_opcodes
```
Comment (matthiasgoergens) [inline ceno_zkvm/src/structs.rs:286]: Cosmetic suggestion: use itertools, we already have it as a dependency anyway.

```diff
diff --git a/ceno_zkvm/src/structs.rs b/ceno_zkvm/src/structs.rs
index b29aba88..5d91eaad 100644
--- a/ceno_zkvm/src/structs.rs
+++ b/ceno_zkvm/src/structs.rs
@@ -9,7 +9,7 @@ use crate::{
 };
 use ceno_emul::StepRecord;
 use ff_ext::ExtensionField;
-use itertools::Itertools;
+use itertools::{chain, Itertools};
 use mpcs::PolynomialCommitmentScheme;
 use multilinear_extensions::{
     mle::DenseMultilinearExtension, virtual_poly_v2::ArcMultilinearExtension,
@@ -283,9 +283,10 @@ impl<E: ExtensionField> ZKVMWitnesses<E> {
 
     /// Iterate opcode circuits, then table circuits, sorted by name.
     pub fn into_iter_sorted(self) -> impl Iterator<Item = (String, RowMajorMatrix<E::BaseField>)> {
-        self.witnesses_opcodes
-            .into_iter()
-            .chain(self.witnesses_tables)
+        chain(
+            self.witnesses_opcodes,
+            self.witnesses_tables,
+        )
     }
 }
```

File: ceno_zkvm/src/structs.rs (hunk 2 above)
Comment (naure) [reply ceno_zkvm/src/structs.rs:208]: There can be other circuits. It is called "opcode" everywhere until we find a more precise name.

File: ceno_zkvm/src/scheme/tests.rs (hunk 1 above)
Comment (matthiasgoergens) [reply ceno_zkvm/src/scheme/tests.rs:124]: Ok, I'll have a look.

File: ceno_zkvm/src/structs.rs (hunk 2 above)
Comment (matthiasgoergens) [reply ceno_zkvm/src/structs.rs:208]: So what's actually in there? The instructions or something else?

File: ceno_zkvm/src/structs.rs (hunk 3 above)
Comment (naure) [reply ceno_zkvm/src/structs.rs]: Done.

**Review by kunxian-xia — APPROVED** [review]

**Pull request merged**